### DISCUSSION TITLE
Reasoning behind pulumi.String/Int

### DISCUSSION THREAD

**Discussion by kokizzu** [pulumi/pulumi#7620] [Pulumi in Practice] [answered]
I looked at the example, but the go version seems overly verbose

```
appLabels := pulumi.StringMap{
   "app": pulumi.String(appName),
}
...
Replicas: pulumi.Int(1)
```

is there any reason behind `pulumi.String` and `pulumi.Int` instead of standard string and integer?

**Comment by cnunciato** [comment]
I've wondered this myself. Tagging @EvanBoyle as someone who may have a good explanation as to this choice (as I'm not finding one in our docs).

**Comment by EvanBoyle** [comment, marked as answer]
The inclusion of the `pulumi`-namespaced types such as `pulumi.Int` and `pulumi.String` is due to a lack of generics support in the Go language when we initially implemented our SDK. To model [Inputs and Output](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/) we usually use generics in other languages. 

In type theory terms, all pulumi resources accept `Input<T>` and `Input<T> = T | Output<T>`. In plain english, if a Pulumi resource accepts a string for a particular property such as `IPAddress`, the you can also supply an output from another resource in your program. This allows you to do both of the following:

```
IPAddress = "x.x.x.x" // assign IPAddress a normal string value (`T`).

IPAddress = EC2Instance.IPAddress // use an output from another resource in my program
                                  // the value of which may not yet be known (`Output<T>`).
```

Since Go does not support generics, we had to create our own wrappers around primitive types to implement Inputs and Outputs in the Go SDK. We'll likely revisit the design of the Go SDK once support for Go generics is out of beta and stabilized.
